File: content/problems/0059-spiral-matrix-ii/solutions/shrinking-boundaries/solution.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> answer(n, vector<int>(n)); int top = 0, bottom = n - 1, left = 0, right = n - 1, value = 1;
        while (top <= bottom && left <= right) { for (int column = left; column <= right; ++column) answer[top][column] = value++; ++top; for (int row = top; row <= bottom; ++row) answer[row][right] = value++; --right; if (top <= bottom) { for (int column = right; column >= left; --column) answer[bottom][column] = value++; --bottom; } if (left <= right) { for (int row = bottom; row >= top; --row) answer[row][left] = value++; ++left; } } return answer;
    }
};

File: content/problems/0059-spiral-matrix-ii/solutions/shrinking-boundaries/Solution.kt
class Solution {
    fun generateMatrix(n: Int): Array<IntArray> {
        val answer = Array(n) { IntArray(n) }; var top = 0; var bottom = n - 1; var left = 0; var right = n - 1; var value = 1
        while (top <= bottom && left <= right) { for (column in left..right) answer[top][column] = value++; top++; for (row in top..bottom) answer[row][right] = value++; right--; if (top <= bottom) { for (column in right downTo left) answer[bottom][column] = value++; bottom-- }; if (left <= right) { for (row in bottom downTo top) answer[row][left] = value++; left++ } }
        return answer
    }
}
